File: workflows/prompt-chaining/code/python/prompt_chaining.py
"""
Prompt Chaining — Sequential LLM calls with optional validation gates.

Each step's output becomes the next step's input. Gates check quality
before proceeding; on failure the chain halts and returns the failing step.

Design doc:  ../../design.md
Overview:    ../../overview.md
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Protocol


# ── LLM interface ─────────────────────────────────────────────────────────────
# Implement this with any provider: OpenAI, Anthropic, local model, etc.

class LLM(Protocol):
    def generate(self, messages: list[dict]) -> str:
        """Send messages and return the assistant's response text."""
        ...


# ── Core types ────────────────────────────────────────────────────────────────

@dataclass
class ChainStep:
    name: str
    prompt_template: str                          # Use {input} for previous output
    validate: Callable[[str], bool] = lambda _: True  # Gate: return False to halt


@dataclass
class ChainResult:
    success: bool
    output: str
    step_outputs: list[str] = field(default_factory=list)
    failed_at: str | None = None  # Step name where validation failed


# ── Implementation ────────────────────────────────────────────────────────────

class PromptChain:
    """
    Runs a sequence of LLM calls where each output feeds into the next step.

    Usage:
        chain = PromptChain(llm, steps=[...])
        result = chain.run("initial input")
    """

    def __init__(self, llm: LLM, steps: list[ChainStep], system: str = ""):
        self.llm = llm
        self.steps = steps
        self.system = system

    def run(self, input: str) -> ChainResult:
        current = input
        step_outputs: list[str] = []

        for step in self.steps:
            messages: list[dict] = []
            if self.system:
                messages.append({"role": "system", "content": self.system})
            messages.append({
                "role": "user",
                "content": step.prompt_template.format(input=current),
            })

            output = self.llm.generate(messages)

            if not step.validate(output):
                return ChainResult(
                    success=False,
                    output=output,
                    step_outputs=step_outputs,
                    failed_at=step.name,
                )

            step_outputs.append(output)
            current = output

        return ChainResult(success=True, output=current, step_outputs=step_outputs)


# ── Example ───────────────────────────────────────────────────────────────────

if __name__ == "__main__":

    class MockLLM:
        """Stub — replace with your actual LLM client."""
        def generate(self, messages: list[dict]) -> str:
            content = messages[-1]["content"]
            return f"[step output from: {content[:50].strip()}]"

    chain = PromptChain(
        llm=MockLLM(),
        system="You are a precise document processor.",
        steps=[
            ChainStep(
                name="extract",
                prompt_template="Extract the key facts from this text:\n\n{input}",
                validate=lambda out: len(out) > 0,
            ),
            ChainStep(
                name="summarize",
                prompt_template="Summarize these facts in 2-3 sentences:\n\n{input}",
            ),
            ChainStep(
                name="format",
                prompt_template="Format this summary as a markdown bullet list:\n\n{input}",
            ),
        ],
    )

    result = chain.run("AI is transforming healthcare through diagnostics and drug discovery.")
    print(f"Success:  {result.success}")
    print(f"Steps:    {len(result.step_outputs)}")
    print(f"Output:\n{result.output}")

File: patterns/prompt-chaining/code/typescript/vercel-ai-sdk/prompt-chaining.ts
/**
 * Prompt Chaining — Vercel AI SDK variant.
 *
 * Pattern: Prompt Chaining (sequential LLM calls where each step's output
 *   feeds the next; optional validator gates halt the chain on failure).
 * Framework: Vercel AI SDK (ai ^4.0.0) + @ai-sdk/anthropic (^1.0.0).
 * Idioms: generateText() per step; a tiny ChainStep type so the same
 *   driver works for any chain shape; validator callbacks return bool to
 *   short-circuit cleanly.
 * Design doc: ../../../design.md (the framework-agnostic
 *   ../../python/prompt_chaining.py runs the same three-step smoke:
 *   extract -> summarize -> format).
 *
 * Install:  pnpm add ai @ai-sdk/anthropic
 *           (or via tsx-on-demand: npx tsx prompt-chaining.ts)
 * Run:      ANTHROPIC_API_KEY=... npx tsx prompt-chaining.ts
 *
 * Note: ESM only.
 * tsconfig snippet: { "module": "ESNext", "moduleResolution": "Bundler", "target": "ES2022" }
 */

import { anthropic } from "@ai-sdk/anthropic";
import { generateText, type LanguageModelV1 } from "ai";

// ── Types ───────────────────────────────────────────────────────────────────

interface ChainStep {
  name: string;
  // The chain feeds `{input}` from the prior step's output into the template.
  promptTemplate: string;
  // Return false to halt the chain at this step.
  validate?: (output: string) => boolean;
}

interface ChainResult {
  success: boolean;
  output: string;
  stepOutputs: string[];
  failedAt: string | null;
}

// ── Driver ──────────────────────────────────────────────────────────────────

class PromptChain {
  constructor(
    private readonly steps: ChainStep[],
    private readonly system = "",
    private readonly model: LanguageModelV1 = anthropic("claude-haiku-4-5"),
  ) {}

  async run(initialInput: string): Promise<ChainResult> {
    let current = initialInput;
    const stepOutputs: string[] = [];

    for (const step of this.steps) {
      const prompt = step.promptTemplate.replace("{input}", current);
      const result = await generateText({
        model: this.model,
        system: this.system || undefined,
        prompt,
      });
      const output = result.text;

      if (step.validate && !step.validate(output)) {
        return {
          success: false,
          output,
          stepOutputs,
          failedAt: step.name,
        };
      }
      stepOutputs.push(output);
      current = output;
    }

    return { success: true, output: current, stepOutputs, failedAt: null };
  }
}

// ── Smoke runner ────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  if (!process.env.ANTHROPIC_API_KEY) {
    console.log("Skipping smoke run — set ANTHROPIC_API_KEY to exercise the chain.");
    return;
  }

  const chain = new PromptChain(
    [
      {
        name: "extract",
        promptTemplate: "Extract the key facts from this text:\n\n{input}",
        validate: (out) => out.length > 0,
      },
      {
        name: "summarize",
        promptTemplate: "Summarize these facts in 2-3 sentences:\n\n{input}",
      },
      {
        name: "format",
        promptTemplate: "Format this summary as a markdown bullet list:\n\n{input}",
      },
    ],
    "You are a precise document processor.",
  );

  const result = await chain.run(
    "AI is transforming healthcare through diagnostics and drug discovery.",
  );

  console.log(`Success: ${result.success}`);
  console.log(`Steps:   ${result.stepOutputs.length}`);
  console.log(`Output:\n${result.output}`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

export { PromptChain };
export type { ChainStep, ChainResult };
